### ISSUE TITLE
twisted.python.randbytes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4468]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#4468|
|Type|enhancement|
|Created|2010-05-31 06:03:45Z|
|Branch|https://github.com/twisted/twisted/tree/randbytes-without-pycrypto-4468|

[twisted/python/randbytes.py](twisted/python/randbytes.py) doesn't seem to offer a lot of functionality over ```os.urandom```, except for two misfeatures:
 1. The option of setting ```fallback=True``` so that it will automatically fallback to insecure random numbers if it can't generate secure ones. This is a terrible idea that nobody should ever do. Behavior like that is one of the causes of the Debian OpenSSL Fiasco, for example. If anyone has a legitimate use case for this behavior I would be fascinated to hear it.
 2. The option of using PyCrypto's ```Crypto.Util.randpool``` if it is present. ```randpool``` is deprecated (according to the announcement at the top of http://www.dlitz.net/software/pycrypto/ on this date, which references [this mailing list thread](http://lists.dlitz.net/pipermail/pycrypto/2008q3/000000.html)). I've looked at ```randpool``` a bit, and it doesn't offer any advantage over ```os.urandom``` that I can see other than the option of falling back to insecure random number generation if it can't generate secure random numbers. ```randpool``` also has a lot of other code to do some useless things about estimating entropy, sampling the current clock, and so on. I can't be sure that I understood its source code because that other stuff made it hard to understand the part I was interested in. The latest git version of PyCrypto comes with this warning: "Deprecated.  Use Random.new() instead. See http://www.pycrypto.org/randpool-broken". However that link gives me a 404 Not Found.
I suspect both of these modules predate the ```os.urandom``` module that was introduced in Python 2.4. Now that Twisted no longer supports Python 2.3, perhaps the time has come to deprecate ```twisted.python.randbytes``` in favor of ```os.urandom```.

<details><summary>Searchable metadata</summary>

```
trac-id__4468 4468
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_randbytes_without_pycrypto_4468 branches/randbytes-without-pycrypto-4468
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__security security
time__1275285825000000 1275285825000000
changetime__1277857676000000 1277857676000000
version__None None
owner__ 
cc__zooko cc__zooko@... cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

How about other platforms, where there is no /dev/urandom?  Does Python automatically make `os.urandom` use whatever platform-specific random source they provide?  How about future platforms which won't provide this?  Or a future Linux replacement for /dev/urandom that's completely superior?

Getting rid of the non-cryptographic fallback may be a good idea, but I don't think completely eliminating this module makes sense.  It's a good API for centralizing our decisions about which random number generator to use.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `closed`|
|-|-|

Feel free to re-open (or open another more specific ticket) if you have further comments.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set status to `reopened`|
|-|-|

Replying to [exarkun](#note_1):
> How about other platforms, where there is no /dev/urandom? 

From the [docs](http://docs.python.org/library/os.html#os.urandom): "If a randomness source is not found, NotImplementedError will be raised."

> Does Python automatically make `os.urandom` use whatever platform-specific random source they provide? 

"On a UNIX-like system this will query /dev/urandom, and on Windows it will use CryptGenRandom."

> How about future platforms which won't provide this?  Or a future Linux replacement for /dev/urandom that's completely superior?
> 
> Getting rid of the non-cryptographic fallback may be a good idea, but I don't think completely eliminating this module makes sense.  It's a good API for centralizing our decisions about which random number generator to use.

I think we should get rid of the fallback and switch to ```os.urandom```, and if that throws a ```NotImplementedError``` fallback to ```Random.new()```?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> I think we should get rid of the fallback and switch to os.urandom, and if that throws a NotImplementedError fallback to Random.new()?

Half of zooko's complaint seems to be that `Crypto.Util.randpool` shouldn't be used at all, so that fallback probably doesn't make sense.

So, specifically, it sounds like we should drop the use of `PyCrypto` entirely and then this ticket will be resolved.

We _could_ also remove the `/dev/urandom` support code, but why should we bother?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @zooko<br>@glyph set status to `new`|
|-|-|

For what it's worth, you have to specifically pass `fallback` to say "I don't care, for this application".  Perhaps calling the method `secureRandom` is misleading, but you can use it for a networked dice-roller or something, so you can run it on your NetBSD set-top box, it's obviously not intended for crypto math.  None of the crypto functions in Twisted use that argument.

I officially do not care about the outcome of this ticket though, because for almost all users on almost all platforms, it already uses `os.urandom`.  I agree with exarkun that it is a good place to centralize decisions about random number generation, but since `os.urandom` itself seems to be doing an okay job of doing that, I won't shed a tear if it goes away either.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29376]) Branching to 'randbytes-without-pycrypto-4468'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/randbytes-without-pycrypto-4468)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli set owner to @exarkun|
|-|-|

Pyflakes reports sys as an unused import in test_randbytes, but other than that it looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [29474]) Merge randbytes-without-pycrypto-4468

Author: exarkun
Reviewer: cyli
Fixes: [#4468](https://github.com/twisted/twisted/issues/4468)

Remove the bogus PyCrypto backend to `twisted.python.randbytes`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
